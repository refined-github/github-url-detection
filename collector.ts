/** @file This needs to be in a separate file so it can bee tree-shaken before being published, while still being importable by tests */

const testableUrls = new Map<string, string[]>();

export function addTests(test: string, urls: string[]): void {
	// @ts-expect-error KISS for Vite
	if (process.env.NODE_ENV !== 'bundling') {
		testableUrls.set(test, urls);
	}
}

export function getTests(detectName: string): string[] {
	if (detectName === 'combinedTestOnly') {
		return ['combinedTestOnly'];
	}

	return testableUrls.get(detectName)?.flatMap(url => url.startsWith('http') ? url : getTests(url) ?? []) ?? [];
}

export function getAllUrls(): Set<string> {
	return new Set<string>([...testableUrls.values()].flat().filter(url => url.startsWith('http')));
}
