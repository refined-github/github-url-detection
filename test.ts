import test from 'ava';
import stripIndent from 'strip-indent';
import {getAllUrls, getTests} from './collector.js';
import * as pageDetect from './index.js';

(globalThis as any).document = {title: ''};
(globalThis as any).location = new URL('https://github.com/');

const allUrls = getAllUrls();

for (const [detectName, detect] of Object.entries(pageDetect)) {
	if (typeof detect !== 'function') {
		continue;
	}

	const validURLs = getTests(detectName);

	if (validURLs[0] === 'combinedTestOnly' || String(detect).startsWith('()=>')) {
		continue;
	}

	test(detectName + ' has tests', t => {
		t.true(
			Array.isArray(validURLs),
			`The function \`${detectName}\` doesn’t have any tests. Set them via \`collect.set()\``,
		);
	});

	if (!Array.isArray(validURLs)) {
		continue;
	}

	for (const url of validURLs) {
		test(`${detectName} ${url.replace('https://github.com', '')}`, t => {
			t.true(
				detect(new URL(url)),
				stripIndent(`
					Is this URL \`${detectName}\`?
						${url.replace('https://github.com', '')}

					• Yes? The \`${detectName}\` test is wrong and should be fixed.
					• No? Remove it from its \`collect.set()\` array.
				`),
			);
		});
	}

	// Skip negatives for this one, it's too long
	if (detectName === 'isRepo') {
		continue;
	}

	for (const url of allUrls) {
		if (!validURLs.includes(url)) {
			test(`${detectName} NO ${url}`, t => {
				t.false(
					detect(new URL(url)),
					stripIndent(`
						Is this URL \`${detectName}\`?
							${url.replace('https://github.com', '')}

						• Yes? Add it to the \`collect.set()\` array.
						• No? The \`${detectName}\` test is wrong and should be fixed.
					`),
				);
			});
		}
	}
}

test('is404', t => {
	document.title = 'Page not found · GitHub';
	t.true(pageDetect.is404());

	document.title = 'examples/404: Page not found examples';
	t.false(pageDetect.is404());

	document.title = 'Dashboard';
	t.false(pageDetect.is404());

	document.title = 'Page not found · Issue #266 · sintaxi/surge · GitHub';
	t.false(pageDetect.is404());
});

test('is500', t => {
	document.title = 'Server Error · GitHub';
	t.true(pageDetect.is500());

	document.title = 'Unicorn! · GitHub';
	t.true(pageDetect.is500());

	document.title = 'examples/500: Server Error examples';
	t.false(pageDetect.is500());

	document.title = 'sindresorhus/unicorn: You can’t tell what doesn’t exist';
	t.false(pageDetect.is500());

	document.title = 'Dashboard';
	t.false(pageDetect.is500());

	document.title = 'Server Error · Issue #266 · sintaxi/surge · GitHub';
	t.false(pageDetect.is500());
});

test('isPRCommit404', t => {
	document.title = 'Commit range not found · Pull Request #3227 · sindresorhus/refined-github';
	location.href = 'https://github.com/sindresorhus/refined-github/pull/3227/commits/32c8a88360a85739f151566eae0225d530ce6a15';
	t.true(pageDetect.isPRCommit404());

	document.title = 'Experiment with `@primer/octicons-react` icons by FloEdelmann · Pull Request #3227 · sindresorhus/refined-github';
	location.href = 'https://github.com/sindresorhus/refined-github/pull/3227/commits/edbdcdd5559a2a8da78abdc7cb0814155713974c';
	t.false(pageDetect.isPRCommit404());

	document.title = 'Commit range not found by SomeContributor · Pull Request #999999 · sindresorhus/refined-github';
	location.href = 'https://github.com/sindresorhus/refined-github/pull/999999/commits/32c8a88360a85739f151566eae0225d530ce6a15';
	t.false(pageDetect.isPRCommit404());
});

test('isPRFile404', t => {
	document.title = 'Commit range not found · Pull Request #789 · sindresorhus/eslint-plugin-unicorn';
	location.href = 'https://github.com/sindresorhus/eslint-plugin-unicorn/pull/789/files/a58b37845f1b2660221de019e4ae6c736feedc26..eed168224d7994652b1d1ff69a5c8cebee223faf';
	t.true(pageDetect.isPRFile404());

	document.title = 'Add `align-repository-header` feature by fregante · Pull Request #3313 · sindresorhus/refined-github';
	location.href = 'https://github.com/sindresorhus/refined-github/pull/3313/files/a14fb2c94eae3ca83a3a97688a171fcc3405524f..fbeeba9825f12b5ded9cd4bb04d5df4b0cf2f2a8';
	t.false(pageDetect.isPRFile404());
});

const {getRepositoryInfo} = pageDetect.utils;
test('getRepositoryInfo', t => {
	const inputTypes = [
		getRepositoryInfo, // Full URL
		(url: string) => getRepositoryInfo(new URL(url).pathname), // Pathname only
		(url: string) => getRepositoryInfo(new URL(url)), // URL object
	];
	for (const getRepositoryInfoAdapter of inputTypes) {
		t.is(getRepositoryInfoAdapter('https://github.com'), undefined);
		t.is(getRepositoryInfoAdapter('https://gist.github.com/'), undefined);
		t.is(getRepositoryInfoAdapter('https://github.com/settings/developers'), undefined);
		t.deepEqual(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: '',
		});
		t.deepEqual(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: '',
		});
		t.deepEqual(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/blame/master/package.json'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'blame/master/package.json',
		});
		t.deepEqual(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/commit/57bf4'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'commit/57bf4',
		});
		t.deepEqual(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/compare/test-branch?quick_pull=0'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'compare/test-branch',
		});
		t.deepEqual(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/tree/master/distribution'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'tree/master/distribution',
		});
		t.deepEqual(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/tree/master/distribution/'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'tree/master/distribution',
		});
	}
});
