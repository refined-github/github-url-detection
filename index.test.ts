import {test, expect} from 'bun:test';
import stripIndent from 'strip-indent';
import collector from './collector.js';
import * as pageDetect from './index.js';

(global as any).document = {title: ''};
(global as any).location = new URL('https://github.com/');

const allUrls = new Set<string>([...collector.values()].flat());

allUrls.delete('combinedTestOnly');

const assert = {
	isTrue(a) {
		expect(a).toBe(true);
	},
	isFalse(a) {
		expect(a).toBe(false);
	},
	equal(a, b) {
		expect(a).toBe(b);
	},
};

for (const [detectName, detect] of Object.entries(pageDetect)) {
	if (typeof detect !== 'function') {
		continue;
	}

	const validURLs = collector.get(detectName);

	if (validURLs === 'combinedTestOnly' || String(detect).startsWith('() =>')) {
		continue;
	}

	test(detectName + ' has tests', () => {
		assert.isTrue(
			Array.isArray(validURLs),
			`The function \`${detectName}\` doesn’t have any tests. Set them via \`collect.set()\``,
		);
	});

	if (!Array.isArray(validURLs)) {
		continue;
	}

	for (const url of validURLs) {
		test(`${detectName} ${url.replace('https://github.com', '')}`, () => {
			assert.isTrue(
				detect(new URL(url)),
				stripIndent(`
					Is this URL \`${detectName}\`?
						${url.replace('https://github.com', '')}

					• Yes? The \`${detectName}\` test is wrong and should be fixed.
					• No? Remove it from its \`collect.set()\` array.
				`),
			);
		});
	}

	// Skip negatives for this one, it's too long
	if (detectName === 'isRepo') {
		continue;
	}

	for (const url of allUrls) {
		if (!validURLs.includes(url)) {
			test(`${detectName} NO ${url}`, () => {
				assert.isFalse(
					detect(new URL(url)),
					stripIndent(`
						Is this URL \`${detectName}\`?
							${url.replace('https://github.com', '')}

						• Yes? Add it to the \`collect.set()\` array.
						• No? The \`${detectName}\` test is wrong and should be fixed.
					`),
				);
			});
		}
	}
}

test('is404', () => {
	document.title = 'Page not found · GitHub';
	assert.isTrue(pageDetect.is404());

	document.title = 'examples/404: Page not found examples';
	assert.isFalse(pageDetect.is404());

	document.title = 'Dashboard';
	assert.isFalse(pageDetect.is404());

	document.title = 'Page not found · Issue #266 · sintaxi/surge · GitHub';
	assert.isFalse(pageDetect.is404());
});

test('is500', () => {
	document.title = 'Server Error · GitHub';
	assert.isTrue(pageDetect.is500());

	document.title = 'Unicorn! · GitHub';
	assert.isTrue(pageDetect.is500());

	document.title = 'examples/500: Server Error examples';
	assert.isFalse(pageDetect.is500());

	document.title = 'sindresorhus/unicorn: You can’t tell what doesn’t exist';
	assert.isFalse(pageDetect.is500());

	document.title = 'Dashboard';
	assert.isFalse(pageDetect.is500());

	document.title = 'Server Error · Issue #266 · sintaxi/surge · GitHub';
	assert.isFalse(pageDetect.is500());
});

test('isPRCommit404', () => {
	document.title = 'Commit range not found · Pull Request #3227 · sindresorhus/refined-github';
	location.href = 'https://github.com/sindresorhus/refined-github/pull/3227/commits/32c8a88360a85739f151566eae0225d530ce6a15';
	assert.isTrue(pageDetect.isPRCommit404());

	document.title = 'Experiment with `@primer/octicons-react` icons by FloEdelmann · Pull Request #3227 · sindresorhus/refined-github';
	location.href = 'https://github.com/sindresorhus/refined-github/pull/3227/commits/edbdcdd5559a2a8da78abdc7cb0814155713974c';
	assert.isFalse(pageDetect.isPRCommit404());

	document.title = 'Commit range not found by SomeContributor · Pull Request #999999 · sindresorhus/refined-github';
	location.href = 'https://github.com/sindresorhus/refined-github/pull/999999/commits/32c8a88360a85739f151566eae0225d530ce6a15';
	assert.isFalse(pageDetect.isPRCommit404());
});

test('isPRFile404', () => {
	document.title = 'Commit range not found · Pull Request #789 · sindresorhus/eslint-plugin-unicorn';
	location.href = 'https://github.com/sindresorhus/eslint-plugin-unicorn/pull/789/files/a58b37845f1b2660221de019e4ae6c736feedc26..eed168224d7994652b1d1ff69a5c8cebee223faf';
	assert.isTrue(pageDetect.isPRFile404());

	document.title = 'Add `align-repository-header` feature by fregante · Pull Request #3313 · sindresorhus/refined-github';
	location.href = 'https://github.com/sindresorhus/refined-github/pull/3313/files/a14fb2c94eae3ca83a3a97688a171fcc3405524f..fbeeba9825f12b5ded9cd4bb04d5df4b0cf2f2a8';
	assert.isFalse(pageDetect.isPRFile404());
});

const {getRepositoryInfo} = pageDetect.utils;
test('getRepositoryInfo', () => {
	const inputTypes = [
		getRepositoryInfo, // Full URL
		(url: string) => getRepositoryInfo(new URL(url).pathname), // Pathname only
		(url: string) => getRepositoryInfo(new URL(url)), // URL object
	];
	for (const getRepositoryInfoAdapter of inputTypes) {
		assert.equal(getRepositoryInfoAdapter('https://github.com'), undefined);
		assert.equal(getRepositoryInfoAdapter('https://gist.github.com/'), undefined);
		assert.equal(getRepositoryInfoAdapter('https://github.com/settings/developers'), undefined);
		assert.equal(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: '',
		});
		assert.equal(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: '',
		});
		assert.equal(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/blame/master/package.json'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'blame/master/package.json',
		});
		assert.equal(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/commit/57bf4'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'commit/57bf4',
		});
		assert.equal(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/compare/test-branch?quick_pull=0'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'compare/test-branch',
		});
		assert.equal(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/tree/master/distribution'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'tree/master/distribution',
		});
		assert.equal(getRepositoryInfoAdapter('https://github.com/refined-github/github-url-detection/tree/master/distribution/'), {
			owner: 'refined-github',
			name: 'github-url-detection',
			nameWithOwner: 'refined-github/github-url-detection',
			path: 'tree/master/distribution',
		});
	}
});
