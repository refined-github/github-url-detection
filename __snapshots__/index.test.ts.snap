// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getRepositoryInfo 1`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "",
  "pathParts": [],
}
`;

exports[`getRepositoryInfo 2`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "",
  "pathParts": [],
}
`;

exports[`getRepositoryInfo 3`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "blame/master/package.json",
  "pathParts": [
    "blame",
    "master",
    "package.json",
  ],
}
`;

exports[`getRepositoryInfo 4`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "commit/57bf4",
  "pathParts": [
    "commit",
    "57bf4",
  ],
}
`;

exports[`getRepositoryInfo 5`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "compare/test-branch",
  "pathParts": [
    "compare",
    "test-branch",
  ],
}
`;

exports[`getRepositoryInfo 6`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "tree/master/distribution",
  "pathParts": [
    "tree",
    "master",
    "distribution",
  ],
}
`;

exports[`getRepositoryInfo 7`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "tree/master/distribution",
  "pathParts": [
    "tree",
    "master",
    "distribution",
  ],
}
`;

exports[`getRepositoryInfo 8`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "",
  "pathParts": [],
}
`;

exports[`getRepositoryInfo 9`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "",
  "pathParts": [],
}
`;

exports[`getRepositoryInfo 10`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "blame/master/package.json",
  "pathParts": [
    "blame",
    "master",
    "package.json",
  ],
}
`;

exports[`getRepositoryInfo 11`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "commit/57bf4",
  "pathParts": [
    "commit",
    "57bf4",
  ],
}
`;

exports[`getRepositoryInfo 12`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "compare/test-branch",
  "pathParts": [
    "compare",
    "test-branch",
  ],
}
`;

exports[`getRepositoryInfo 13`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "tree/master/distribution",
  "pathParts": [
    "tree",
    "master",
    "distribution",
  ],
}
`;

exports[`getRepositoryInfo 14`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "tree/master/distribution",
  "pathParts": [
    "tree",
    "master",
    "distribution",
  ],
}
`;

exports[`getRepositoryInfo 15`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "",
  "pathParts": [],
}
`;

exports[`getRepositoryInfo 16`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "",
  "pathParts": [],
}
`;

exports[`getRepositoryInfo 17`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "blame/master/package.json",
  "pathParts": [
    "blame",
    "master",
    "package.json",
  ],
}
`;

exports[`getRepositoryInfo 18`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "commit/57bf4",
  "pathParts": [
    "commit",
    "57bf4",
  ],
}
`;

exports[`getRepositoryInfo 19`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "compare/test-branch",
  "pathParts": [
    "compare",
    "test-branch",
  ],
}
`;

exports[`getRepositoryInfo 20`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "tree/master/distribution",
  "pathParts": [
    "tree",
    "master",
    "distribution",
  ],
}
`;

exports[`getRepositoryInfo 21`] = `
{
  "name": "github-url-detection",
  "nameWithOwner": "refined-github/github-url-detection",
  "owner": "refined-github",
  "path": "tree/master/distribution",
  "pathParts": [
    "tree",
    "master",
    "distribution",
  ],
}
`;
